store/primary/primary.go

```diff
@@ -69,47 +76,87 @@ func (s *rtStorage) Get(c cid.Cid) ([]store.IndexEntry, bool) {
 // changes consider using a map.
 func (s *rtStorage) Put(c cid.Cid, provID peer.ID, pieceID cid.Cid) error {
 	in := store.IndexEntry{ProvID: provID, PieceID: pieceID}
-	// Check size of primary storage for eviction purposes.
-	s.checkSize()
+
 	// Get multihash from cid
 	k := string(c.Hash())
-	s.primary.lk.Lock()
 
-	old, found := s.primary.tree.Get(k)
-	// If found it means there is already a value there.
-	if found {
-		// Check if we are trying to put a duplicate entry
-		// NOTE: If we end up having a lot of entries for the
-		// same CID we may choose to change IndexEntry to a map[peer.ID]pieceID
-		// to speed-up this lookup. Don't think is the case right now.
-		if !duplicateEntry(in, old.([]store.IndexEntry)) {
-			// If not duplicate entry, append to the end
-			s.primary.tree.Put(k, append(old.([]store.IndexEntry), in))
-		}
+	s.lockBranch(k)
+
+	if !s.put(k, in) {
+		s.unlockBranch(k)
+		return nil
+	}
+
+	if s.current.Len() >= s.sizeLimit {
+		s.unlockBranch(k)
+		s.rotateCache()
 	} else {
-		s.primary.tree.Put(k, []store.IndexEntry{in})
+		s.unlockBranch(k)
 	}
-	s.primary.lk.Unlock()
 
-	// NOTE: Insert in the radix-tree used doesn't return any error.
-	// This may change in the future so keeping an error as output in the signature
-	// for now.
 	return nil
 }
 
+func (s *rtStorage) put(k string, entry store.IndexEntry) bool {
+	// Get from current or previous cache
+	old, found := s.get(k)
+	// If found it means there is already a value there.
+	// Check if we are trying to put a duplicate entry
+	// NOTE: If we end up having a lot of entries for the
+	// same CID we may choose to change IndexEntry to a map[peer.ID]pieceID
+	// to speed-up this lookup. Don't think is the case right now.
+	if found && duplicateEntry(entry, old) {
+		return false
+	}
+
+	s.current.Put(k, append(old, entry))
+	return true
+}
+
+func (s *rtStorage) rotateCache() {
+	// Acquire all locks
+	for i := range s.branchLocks {
+		s.branchLocks[i].Lock()
+	}
+
+	if s.current.Len() >= s.sizeLimit {
+		s.previous, s.current = s.current, radixtree.New()
+	}
+
+	// Release all locks
+	for i := range s.branchLocks {
+		s.branchLocks[i].Unlock()
+	}
+}
+
 // PutMany puts store.IndexEntry information in several CIDs.
 // This is usually triggered when a bulk update for a providerID-pieceID
 // arrives.
 func (s *rtStorage) PutMany(cids []cid.Cid, provID peer.ID, pieceID cid.Cid) error {
+	in := store.IndexEntry{ProvID: provID, PieceID: pieceID}
+	var addedEntry bool
+
 	for i := range cids {
-		err := s.Put(cids[i], provID, pieceID)
-		if err != nil {
-			return err
+		k := string(cids[i].Hash())
+		s.lockBranch(k)
+
+		if s.put(k, in) {
+			addedEntry = true
+		}
+		s.unlockBranch(k)
+	}
+
+	if addedEntry {
```

Comment:
Here we are checking the size of the cache after putting the batch, right? Shouldn't we check it after each individual put better? If the batch is too large we may not realize that we are at capacity until we've finished putting every key.